lib/rust/flexer/src/generate.rs

```diff
@@ -0,0 +1,526 @@
+                        self.#rule(reader);
+                        self.bookmarks.bookmark(matched_bookmark,reader);
+                        StageStatus::ExitSuccess
+                    }})
+                }
+            }
+        }
+    } else {
+        let target_state_has_no_rule = match maybe_state {
+            Some(state) => if !dfa.has_rule_for(target_state) {
+                dfa.callbacks[target_state.id] = Some(state.clone());
+                has_overlaps.insert(target_state.id,true);
+                true
+            } else {
+                false
+            },
+            None => false
+        };
+
+        let state_id = Literal::usize_unsuffixed(target_state.id);
+        let ret:Expr = parse_quote! {
+            StageStatus::ContinueWith(#state_id.into())
+        };
+
+        if target_state_has_no_rule && !rules_overlap {
+            Ok(parse_quote! {{
+                let rule_bookmark = self.bookmarks.rule_bookmark;
+                self.bookmarks.bookmark(rule_bookmark,reader);
+                #ret
+            }})
+        } else {
+            Ok(parse_quote! {{
+                #ret
+            }})
+        }
+    }
+}
+
+/// Generate the dispatch function for a given lexer state.
+///
+/// This dispatch function is responsible for dispatching based on the sub-state of any given lexer
+/// state, and is the main part of implementing the actual lexer transitions.
+pub fn dispatch_in_state(dfa:&DFA, id:usize) -> Result<ImplItem,GenError> {
+    let dispatch_name:Ident = str_to_ident(format!("dispatch_in_state_{}",id))?;
+    let state_names  = dfa.links.row_indices().map(|ix| (ix, name_for_step(id, ix))).collect_vec();
+    let mut branches = Vec::with_capacity(state_names.len());
+    for (ix,name) in state_names.into_iter() {
+        let literal = Literal::usize_unsuffixed(ix);
+        let arm:Arm = parse_quote! {
+            #literal => self.#name(reader),
+        };
+        branches.push(arm);
+    }
+
+    let pattern_match:ExprMatch = parse_quote! {
+        match new_state_index.into() {
+            #(#branches)*
+            _ => unreachable_panic!("Unreachable state reached in lexer.")
+        }
+    };
+    let func:ImplItem = parse_quote! {
+        fn #dispatch_name<R:LazyReader>
+        ( &mut self
+        , new_state_index:SubStateId
+        , reader:&mut R
+        ) -> StageStatus {
+            #pattern_match
+        }
+    };
+
+    Ok(func)
+}
+
+/// Generate a name for a given step function.
+pub fn name_for_step(in_state:usize, to_state:usize) -> Ident {
+    let name_str = format!("state_{}_to_{}",in_state,to_state);
+    parse_str(name_str.as_str()).expect("Impossible to not be a valid identifier.")
+}
+
+/// Generate an executable rule function for a given lexer state.
+pub fn rule_for_state(state:&State) -> Result<ImplItem,GenError> {
+    match &state.name {
+        None => unreachable_panic!("Should never reach here."),
```

Comment:
unreachable tell sthat already. I believe the string should explain shortly why it should not be here as from the code context its not clear to me